### PR TITLE
[ASC-ShareCode-7]restore run profile configuration on reopening deployment dialog

### PR DESCRIPTION
**Pull request by wangmingliang-ms**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfigurationPanel.java [hunk 1]
```diff
@@ -85,11 +90,13 @@ private void onClusterChanger(final ItemEvent e) {
 
     public void setData(SpringCloudAppConfig appConfig) {
         final SpringCloudDeploymentConfig deploymentConfig = appConfig.getDeployment();
-        this.cbCPU.setSelectedItem(Optional.ofNullable(deploymentConfig.getCpu()).orElse(1));
-        this.cbMemory.setSelectedItem(Optional.ofNullable(deploymentConfig.getMemoryInGB()).orElse(1));
-        this.cbInstanceCount.setSelectedItem(Optional.ofNullable(deploymentConfig.getInstanceCount()).orElse(1));
+        this.cbCPU.setSelectedItem(Optional.ofNullable(deploymentConfig.getCpu()).map(String::valueOf).orElse("1"));
+        this.cbMemory.setSelectedItem(Optional.ofNullable(deploymentConfig.getMemoryInGB()).map(String::valueOf).orElse("1"));
+        this.cbInstanceCount.setSelectedItem(Optional.ofNullable(deploymentConfig.getInstanceCount()).map(String::valueOf).orElse("1"));
         this.textJvmOptions.setText(deploymentConfig.getJvmOptions());
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfigurationPanel.java:96]: why add FakeValue

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java [hunk 2]
```diff
@@ -53,26 +52,28 @@ public SpringCloudDeploymentConfiguration(@NotNull Project project, @NotNull Con
...
-    protected SpringCloudDeploymentConfiguration(@NotNull SpringCloudDeploymentConfiguration source) {
-        super(source.getProject(), source.getFactory(), source.getName());
-        this.appConfig = JsonUtils.deepCopyWithJson(source.appConfig);
-    }
-
     @Override
     public void readExternal(Element element) throws InvalidDataException {
         super.readExternal(element);
-        XmlSerializer.deserializeInto(this.appConfig, element);
+        final Element appConfigElement = element.getChild("SpringCloudAppConfig");
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java:58]: extract the encode/decode logic code in readExternal and writeExternal to another class and write a unit test against it.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java [hunk 3]
```diff
@@ -53,26 +52,28 @@ public SpringCloudDeploymentConfiguration(@NotNull Project project, @NotNull Con
...
-        XmlSerializer.serializeInto(this.appConfig, element, new SerializationFilterBase() {
-            @Override
-            protected boolean accepts(@NotNull Accessor accessor, @NotNull Object bean, @Nullable Object beanValue) {
-                return !"password".equalsIgnoreCase(accessor.getName());
-            }
-        });
+        final Element appConfigElement = XmlSerializer.serialize(this.appConfig, (accessor, o) -> !"artifact".equalsIgnoreCase(accessor.getName()));
+        final IArtifact artifact = this.appConfig.getDeployment().getArtifact();
+        Optional.ofNullable(this.appConfig)
+            .map(config -> XmlSerializer.serialize(config, (accessor, o) -> !"artifact".equalsIgnoreCase(accessor.getName())))
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java:70]: StringUtils.equalsIgnoreCase

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java [hunk 4]
```diff
@@ -53,26 +52,28 @@ public SpringCloudDeploymentConfiguration(@NotNull Project project, @NotNull Con
...
+            .ifPresent(element::addContent);
+        Optional.ofNullable(this.appConfig)
+            .map(config -> config.getDeployment().getArtifact())
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java:73]: config.getDeployment() is non-null?

Comment (wangmingliang-ms) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java:73]: yes

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java (hunk 3 above)
Comment (wangmingliang-ms) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfiguration.java:70]: it's safe to call `String.equalsIgnoreCase` here

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfigurationPanel.java (hunk 1 above)
Comment (wangmingliang-ms) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/springcloud/deplolyment/SpringCloudDeploymentConfigurationPanel.java:96]: to set selected item/value without constructing a complex instance. 
renamed to ItemReference.

**Review by Flanker32 — APPROVED** [review]

**Pull request merged**